---
title: "Corrected Readme"
author: "Manuela Chaves"
date: "2024-03-21"
output: html_document
---


During this R assignment, two raw genomic data files (“fang _et_al_genotypes.txt” and “snp_position.txt”) were subjected to a file inspection process through Rstudio, to describe and understand the kind of data we are dealing with, as well as to pre-process the data for further analysis.

## Files:
fang_et_al_genotypes.txt: a published SNP data set including maize, teosinte (i.e., wild maize), and Tripsacum (a close outgroup to the genus Zea) individuals.

snp_position.txt: an additional data file that includes the SNP id (first column), chromosome location (third column), nucleotide location (fourth column) and other information for the SNPs genotyped in the fang_et_al_genotypes.txt file.

## Note:
1. This Readme file implemented corrections from reviewers.Go to "Readme.rmd" to see the first version of the project.
2. The created_files folder contain all created files in this R project in case you need them.

# Part 1. Data inspection


## Reading .txt files:

* The following command chunk will run the initial files, which are contained in the same repository folder.

```{r load libraries}
# Load required libraries
install.packages("ggplot2")
install.packages("tidyverse")
library(readr)
library(data.table)
library(dplyr)
library(ggplot2)
library(tidyr)

```

```{r read txt}
snp <- read_tsv("snp_position.txt", show_col_types = FALSE)
fang <- read_tsv("fang_et_al_genotypes.txt", show_col_types = FALSE)
```


## Attributes of the Fang et al & SNP files

* The following chunk shows the commands I used to inspect the data:

```{r inspection}
#Inspecting the Fang file

dim(fang) #dimensions of the file
str(fang[, 1:5]) #structure of the file for the first 5 columns
#summary(fang[, 1:5]) # statistic summary 
names(fang) #shows names of all columns in file
head(fang) #shows first lines of file
tail(fang) #shows last lines of file
na_fang <- grepl("?/?", fang, ignore.case = TRUE) #check for missing values

#Inspecting the SNP file
dim(snp)
str(snp)
summary(snp)
names(snp)
head(snp)
tail(snp)
na_snp <- grepl("NA", snp, ignore.case = TRUE) 
```

### Assessing inspection

1. Fang:
* Number of rows = 2782 and columns = 986
* Data structure shows character type (col_character). 
* The summary command is not very useful for the Fang file since it recognizes the data as characters. That is why I commented out this line.
* Tail and head will display part of the data nicely organized in a printed data frame (tibble)
* We can check the names of all the columns in the file with "names()"
* To check for missing values in the Fang file we could use $ colSums(is.na(fang)), which will return a logical matrix of the same dimensions of the file, and sums the TRUE values in each column, TRUE meaning NA. This command returns all 0, meaning all FALSE (not NA). However, we know from the previous UNIX assignment, that the missing data in Fang is delimited as "?/?", in which case we can use na_fang <- grepl("?/?", fang, ignore.case = TRUE), which will create a data frame with insensitive search, meaning any combination of the missing value. This time we get that there is at least 1 missing data in every column (all TRUE values).


2. SNP:
* Number of rows = 983 and columns = 15
* Data structure includes characters (col_character) and numeric (col_double) types.
* Summary shows statistics for numeric columns.
* Tail and head will display part of the data nicely organized in a printed data frame (tibble)
* We can check the names of all the columns in the file with names()
* Missing values in the SNP file are delimited by NA, in this case we can use the colSums(is.na(snp)) command, and it will return quickly the count of missing data and the column. However this command is sensitive, meaning that if there is any NA value in different notation, the command will not count it. Thus, we can use the na_snp <- grepl("NA", snp, ignore.case = TRUE) to be sure of the missing data results.


# Part 2. Data processing

## Extracting data

* We need to get columns 1, 2, and 3 for SNP_ID, Chromosome and Position respectively. 

```{r data extraction}
# Extract the indices of columns matching the pattern, and order by SNP ID.
extracted_columns <- snp[, c(1, 3, 4)]
extracted_columns <- arrange(extracted_columns, SNP_ID)
```

* Extract specifically the Maize (ZMM) and Teosinte (ZMP) groups along with the SNP_ID (in columns) of the Fang file.

```{r data extraction}
#Filter the rows with ZMM maize groups
zmm <- filter(fang, startsWith(Group, 'ZMM')) 

#transpose file, add back column and row names. 
t_zmm <- transpose(zmm)
rownames(t_zmm) <- colnames(zmm)
colnames(t_zmm) <- rownames(zmm)
setDT(t_zmm, keep.rownames = TRUE)
colnames(t_zmm)[colnames(t_zmm) == 'rn'] <- 'SNP_ID'

#Teosinte 
zmp <- filter(fang, startsWith(Group, 'ZMP')) 
t_zmp <- transpose(zmp)
rownames(t_zmp) <- colnames(zmp)
colnames(t_zmp) <- rownames(zmp)
setDT(t_zmp, keep.rownames = TRUE)
colnames(t_zmp)[colnames(t_zmp) == 'rn'] <- 'SNP_ID'

```

## Merging data files

* Now we need to merge the files snp_columns.txt with the transpose data for Maize and Teosinte.

```{r data merging}
# Merging Maize file
maize_snp <- merge(extracted_columns, t_zmm, by = 'SNP_ID')
# Merging Teosinte file
teosinte_snp <- merge(extracted_columns, t_zmp, by = 'SNP_ID')
```

Right now we count with a merged file organized as we need it according to the assignment instructions: 
column 1 = SNP-ID
column 2 = Chromosome
column 3 = Position
column 4:978 = genotype 

## Creating the files for the assignment:

The assignment is requesting us to create the following files:

### Maize files
#### 1. 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

* The data is naturally encoded ? for missing data.
* For these files we need to order the SNP_ID according to Position in increasing order.
* The Position column contains non-numerical values (unknown, multiple and empty) which will interfere when trying to order the data based on increasing/decreasing values. Therefore, I am taking those values out and creating a clean version (num_open_merged_zmm), which will be used to obtain the file that we need. It uses the grepl() function to identify rows where the Position column consists entirely of digits.
* This chunk first converts the Position column of open_merged_zmm dataframe to numeric, then converts it to numeric, later it orders it based on the values of the Position column, and finally prints the ordered numeric values.
* In order to check if the ordering worked, we need to make sure our data is numeric, that is why I used twice the as.numeric command. 
* Finally, We can check if the data is organized in increasing Position (column 3) by converting the matrix array into a data frame (if needed) and printing a TRUE result if the data is actually increasing. Result printed TRUE, meaning it is increasing.

```{r increasing position}
# Let's extract the non-numerical values from the 'Position' column 
num_open_merged_zmm <- maize_snp[grepl("^\\d+$", maize_snp$Position), ]

# Convert 'Position' column into numeric
convnum_open_merged_zmm <- as.numeric(num_open_merged_zmm$Position)

# Order the numeric values by the 'Position' column
inc_zmm <- num_open_merged_zmm[order(convnum_open_merged_zmm), ]

# Let's make sure the position column is numeric 
num_inc_zmm <- as.numeric(inc_zmm$Position)

# Check if the 'Position' column is in ascending order
is_ordered <- all(diff(num_inc_zmm) >= 0)

# Print the result
print(is_ordered)
```
* We now need to create 1 file for each chromosome (1:10) for the organized file. We can do that with a for loop command that will store automatically the .txt files for the corresponding chromosome. 

```{r chromosome loop}
# Get unique chromosome numbers
chromosomeszmm <- unique(inc_zmm$Chromosome)

# Loop through each chromosome
for (chr in chromosomeszmm) {
  # Filter data for the current chromosome
  chr_data <- inc_zmm[inc_zmm$Chromosome == chr, ]
  
  # Write the filtered data to a separate file
  file_name <- paste0("maize_increasing_chr_", chr, ".txt")
  write.table(chr_data, file = file_name, sep = "\t", quote = FALSE, row.names = FALSE)
}
```


#### 2. 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
* For these files we need to order the SNP_ID according to Position in decreasing order. We can use the numeric clean file that we got in the previous item (convnum_open_merged_zmm).
* Another reason why we convert the non-numeric values to numeric is that we could not use the unary operator - (negation) on the data. But once the data is converted to numeric there is no problem.
* When checking for decreasing order, we get FALSE, meaning the order is not ascending.

```{r decreasing position}
# Order the original data frame by the numeric values in 'Position' column in decreasing order
dec_zmm <- num_open_merged_zmm[order(-convnum_open_merged_zmm), ]

# Let's make sure the position column is numeric 
num_dec_zmm <- as.numeric(dec_zmm$Position)

# Check if the 'Position' column is in decreasing order
is_ordered_dec <- all(diff(num_dec_zmm) >= 0)

# Print the result
print(is_ordered_dec)

```

* Now we need to change the encoded symbol ?/? for -/-.We can do it with lapply command. However, this transforms the data into a list and we need to convert back to data frame. 
* Then we can create 1 file for each chromosome (1:10) for the organized file as we did previously.

```{r replace symbols}
# Using lapply to replace "?" with "-" for the entire dataframe
encoded_dec_zmm <- lapply(dec_zmm, function(x) gsub("\\?", "-", x))

# Convert the list back to a dataframe
encoded_dec_zmm_df <- as.data.frame(encoded_dec_zmm)

# Now proceed with writing data to files for each chromosome
# Get unique chromosome numbers
chromosomeszmm_dec <- unique(encoded_dec_zmm_df$Chromosome)

# Loop through each chromosome
for (chr in chromosomeszmm_dec) {
  chr_data <- encoded_dec_zmm_df[encoded_dec_zmm_df$Chromosome == chr, ]
  file_name <- paste0("maize_decreasing_chr_", chr, ".txt")
  write.table(chr_data, file = file_name, sep = "\t", quote = FALSE, row.names = FALSE)
}
```

### Teosinte files
#### 3. 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

* For these files we just need to replace the commands we used for item 1 with the respective data from Teosinte (zmp)
* When checking for increasing order result is TRUE.
* As a reminder, the data is naturally encoded with "?" for missing values. 

```{r increasing order}
# Let's extract the non-numerical values from the 'Position' column 
num_open_merged_zmp <- teosinte_snp[grepl("^\\d+$", teosinte_snp$Position), ]

# Convert 'Position' column into numeric
convnum_open_merged_zmp <- as.numeric(num_open_merged_zmp$Position)

# Order the numeric values by the 'Position' column
inc_zmp <- num_open_merged_zmp[order(convnum_open_merged_zmp), ]

# Let's make sure the position column is numeric 
num_inc_zmp <- as.numeric(inc_zmp$Position)

# Check if the 'Position' column is in ascending order
is_ordered <- all(diff(num_inc_zmp) >= 0)

# Print the result
print(is_ordered)

```

```{r chromosome loop}

# Get unique chromosome numbers
chromosomeszmp <- unique(inc_zmp$Chromosome)

# Loop through each chromosome
for (chr in chromosomeszmp) {
  # Filter data for the current chromosome
  chr_data <- inc_zmp[inc_zmp$Chromosome == chr, ]
  
  # Write the filtered data to a separate file
  file_name <- paste0("teosinte_increasing_chr_", chr, ".txt")
  write.table(chr_data, file = file_name, sep = "\t", quote = FALSE, row.names = FALSE)
}
```

#### 4. 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
* For these files we just need to replace the commands we used for item 2 with the respective data from Teosinte (zmp)
* When checking for increasing order result is FALSE.

```{r decreasing position}
# Order the original data frame by the numeric values in 'Position' column in decreasing order
dec_zmp <- num_open_merged_zmp[order(-convnum_open_merged_zmp), ]

# Let's make sure the position column is numeric 
num_dec_zmp <- as.numeric(dec_zmp$Position)

# Check if the 'Position' column is in decreasing order
is_ordered_dec <- all(diff(num_dec_zmp) >= 0)

# Print the result
print(is_ordered_dec)

```

```{r chromosome loop}
# Using lapply to replace "?" with "-" for the entire dataframe
encoded_dec_zmp <- lapply(dec_zmp, function(x) gsub("\\?", "-", x))

# Convert the list back to a dataframe
encoded_dec_zmp_df <- as.data.frame(encoded_dec_zmp)

# Now proceed with writing data to files for each chromosome
# Get unique chromosome numbers
chromosomeszmp_dec <- unique(encoded_dec_zmp_df$Chromosome)

# Loop through each chromosome
for (chr in chromosomeszmp_dec) {
  chr_data <- encoded_dec_zmp_df[encoded_dec_zmp_df$Chromosome == chr, ]
  file_name <- paste0("teosinte_decreasing_chr_", chr, ".txt")
  write.table(chr_data, file = file_name, sep = "\t", quote = FALSE, row.names = FALSE)
}

# If you want to check the files you can use
# teosinte_dec_1 = read.table("C:/Users/mchavesm/Box/Rass/R_assignment/teosinte_decreasing_chr_1.txt", header = TRUE)

```

# Conclusion:
We have successfully process our 40 files.

# Part 3. Data visualization

## SNPs per chromosome

1. What is the distribution of SNPs on and accross chromosomes?

* Count the SNPs present in both Maize and Teosinte groups in each chromosome
```{r snp count}
zmm_counts <- num_open_merged_zmm %>% count(Chromosome, sort = TRUE)
zmp_counts <- num_open_merged_zmp %>% count(Chromosome, sort = TRUE)

```

Now, create the distribution plot. 

```{r snp count}
# Maize
ggplot() + geom_col(data = zmm_counts, mapping = aes(x=Chromosome, y=n)) + labs(x = "Chromosome", y = "Number of SNPs", fill = "Chromosome")

# Teosinte
ggplot() + geom_col(data = zmp_counts, mapping = aes(x=Chromosome, y=n)) + labs(x = "Chromosome", y = "Number of SNPs", fill = "Chromosome")
```

2. Are there more SNP positions in maize or teosinte individuals?

Visually, both Maize and Teosinte groups contain the same amount of SNPs in their chromosomes.

## Missing data and amount of heterozygosity

1. What is the proportion of homozygous and heterozygous sites as well as missing data in each sample and each group?

To answer this question, first we need to search and count both homozygous and heterozygous sites in our files. We will create a new data frame for those sites, as well as for the missing data.

```{r snp count}

#Maize homozygous
homozygous_zmm <- data.frame(rowSums(num_open_merged_zmm== "A/A" | num_open_merged_zmm=="T/T" | num_open_merged_zmm== "C/C" | num_open_merged_zmm== "G/G", na.rm = FALSE))

#Maize heterozygous
heterozygous_zmm <- data.frame(data.frame(rowSums(num_open_merged_zmm== "A/T" | num_open_merged_zmm=="A/C" | num_open_merged_zmm== "A/G" | num_open_merged_zmm== "T/A" | num_open_merged_zmm== "T/C" | num_open_merged_zmm== "T/G" | num_open_merged_zmm== "C/A" | num_open_merged_zmm== "C/T" | num_open_merged_zmm== "C/G" | num_open_merged_zmm== "G/A" | num_open_merged_zmm== "G/T" | num_open_merged_zmm== "G/C", na.rm = FALSE)))

#Maize missing data
missing_zmm <- data.frame(rowSums(num_open_merged_zmm== "?/?", na.rm = FALSE))

#Teosinte homozygous
homozygous_zmp <- data.frame(rowSums(num_open_merged_zmp == "A/A" | num_open_merged_zmp =="T/T" | num_open_merged_zmp == "C/C" | num_open_merged_zmp == "G/G", na.rm = FALSE))

#Teosinte heterozygous
heterozygous_zmp <- data.frame(data.frame(rowSums(num_open_merged_zmp == "A/T" | num_open_merged_zmp =="A/C" | num_open_merged_zmp == "A/G" | num_open_merged_zmp == "T/A" | num_open_merged_zmp == "T/C" | num_open_merged_zmp == "T/G" | num_open_merged_zmp == "C/A" | num_open_merged_zmp == "C/T" | num_open_merged_zmp == "C/G" | num_open_merged_zmp == "G/A" | num_open_merged_zmp == "G/T" | num_open_merged_zmp == "G/C", na.rm = FALSE)))

#Teosintemissing data
missing_zmp <- data.frame(rowSums(num_open_merged_zmp == "?/?", na.rm = FALSE))

```

Then, generate the proportion of homozygous/heterozygous for each group.
```{r snp count}

# Calculate proportion homozygous/heterozygous
p_homo_hetero <- data.frame(homozygous_zmm/heterozygous_zmm)

#remove infinite where division by zero.
p_homo_hetero[sapply(p_homo_hetero, is.infinite)] <- NA

#ADD SNP number due to needing numerical data to actually graph.
SNP_number <- c(1:939)

#Combine all data into one dataframe and then rename the columns as needed.
zmm_proportion <- data.frame(SNP_number, num_open_merged_zmm[1:3], homozygous_zmm, heterozygous_zmm, missing_zmm, p_homo_hetero)

colnames(zmm_proportion)[1] <- 'SNP_number'
colnames(zmm_proportion)[4] <- 'homozygous'
colnames(zmm_proportion)[5] <- 'heterozygous'
colnames(zmm_proportion)[6] <- 'missing'
colnames(zmm_proportion)[7] <- 'homozygous/heterozygous'

# Same for Teosinte
# Calculate proportion homozygous/heterozygous
p_homo_hetero_zmp <- data.frame(homozygous_zmp/heterozygous_zmp)

#remove infinite where division by zero.
p_homo_hetero_zmp[sapply(p_homo_hetero_zmp, is.infinite)] <- NA

#ADD SNP number due to needing numerical data to actually graph.
SNP_number_zmp <- c(1:939)

#Combine all data into one dataframe and then rename the columns as needed.
zmp_proportion <- data.frame(SNP_number_zmp, num_open_merged_zmp[1:3], homozygous_zmp, heterozygous_zmp, missing_zmp, p_homo_hetero_zmp)

colnames(zmp_proportion)[1] <- 'SNP_number'
colnames(zmp_proportion)[4] <- 'homozygous'
colnames(zmp_proportion)[5] <- 'heterozygous'
colnames(zmp_proportion)[6] <- 'missing'
colnames(zmp_proportion)[7] <- 'homozygous/heterozygous'

```


Now, graphic the homozygous, heterozygous, missing and proportion of sites
```{r snp count}
# Maize
#graph homozygous
ggplot() + geom_bar(zmm_proportion, mapping = aes(x= SNP_number, y= homozygous), stat = "identity") + scale_x_continuous() + labs(title = "Amount of Homozygous SNP Sites in Maize", x = "SNP ID", y = "Amount of Homozygous SNPs")

#graph heterozygous maize
ggplot() + geom_bar(zmm_proportion, mapping = aes(x= SNP_number, y= heterozygous), stat = "identity") + scale_x_continuous() + labs(title = "Amount of Heterozygous SNP Sites in Maize", x = "SNP ID", y = "Amount of Heterozygous SNPs")

#graph missing for maize
ggplot() + geom_bar(zmm_proportion, mapping = aes(x= SNP_number, y= missing), stat = "identity") + scale_x_continuous() + labs(title = "Amount of Missing SNP Sites in Maize", x = "SNP ID", y = "Amount of Missing SNPs")

#plot ratio maize
ggplot() + geom_bar(zmm_proportion, mapping = aes(x= SNP_number, y= "homozygous/heterozysgous"), stat = "identity") + scale_x_continuous() + labs(title = "Proportion of homozygous and heterozygous in Maize", x = "SNP ID", y = "Proportion of homozygous and heterozygous")


# Teosinte
#graph homozygous
ggplot() + geom_bar(zmp_proportion, mapping = aes(x= SNP_number, y= homozygous), stat = "identity") + scale_x_continuous() + labs(title = "Amount of Homozygous SNP Sites in Maize", x = "SNP ID", y = "Amount of Homozygous SNPs")

#graph heterozygous maize
ggplot() + geom_bar(zmp_proportion, mapping = aes(x= SNP_number, y= heterozygous), stat = "identity") + scale_x_continuous() + labs(title = "Amount of Heterozygous SNP Sites in Maize", x = "SNP ID", y = "Amount of Heterozygous SNPs")

#graph missing for maize
ggplot() + geom_bar(zmp_proportion, mapping = aes(x= SNP_number, y= missing), stat = "identity") + scale_x_continuous() + labs(title = "Amount of Missing SNP Sites in Maize", x = "SNP ID", y = "Amount of Missing SNPs")

#plot ratio maize
ggplot() + geom_bar(zmp_proportion, mapping = aes(x= SNP_number, y= "homozygous/heterozysgous"), stat = "identity") + scale_x_continuous() + labs(title = "Proportion of homozygous and heterozygous in Teosinte", x = "SNP ID", y = "Proportion of homozygous and heterozygous")

```

# My visualization

For my own visualization I would like to know if there are repeated SNPs in the different positions. For that, I will look at the presence of duplicate SNPs.

```{r visualization}
# Maize
# Check for duplicated SNP numbers
duplicated_SNPs_zmm <- zmm_proportion$SNP_number[duplicated(zmm_proportion$SNP_number)]

# Print or inspect the duplicated SNP numbers
if (length(duplicated_SNPs_zmm) > 0) {
  print(paste("Duplicated SNP numbers:", toString(duplicated_SNPs_zmm)))
} else {
  print("No duplicated SNP numbers found.")
}
library(ggplot2)

# Count the frequency of each SNP number
snp_frequency <- table(zmm_proportion$SNP_number)

# Create a data frame for plotting
snp_plot_data <- data.frame(SNP_number = as.numeric(names(snp_frequency)),
                            Frequency = as.numeric(snp_frequency))

# Create a scatter plot
ggplot(snp_plot_data, aes(x = Frequency, y = SNP_number)) +
  geom_point(size = 2) +
  labs(title = "Frequency of SNP Numbers",
       x = "Frequency",
       y = "SNP Number") +
  theme_minimal()

```

```{r visualization}
# Teosinte

# Check for duplicated SNP numbers
duplicated_SNPs_zmp <- zmp_proportion$SNP_number_zmp[duplicated(zmp_proportion$SNP_number_zmp)]

# Print or inspect the duplicated SNP numbers
if (length(duplicated_SNPs_zmp) > 0) {
  print(paste("Duplicated SNP numbers:", toString(duplicated_SNPs_zmp)))
} else {
  print("No duplicated SNP numbers found.")
}
library(ggplot2)

# Count the frequency of each SNP number
snp_frequency <- table(zmp_proportion$SNP_number_zmp)

# Create a data frame for plotting
snp_plot_data_zmp <- data.frame(SNP_number_zmp = as.numeric(names(snp_frequency)),
                            Frequency = as.numeric(snp_frequency))

# Create a scatter plot
ggplot(snp_plot_data, aes(x = Frequency, y = SNP_number_zmp)) +
  geom_point(size = 2) +
  labs(title = "Frequency of SNP Numbers",
       x = "Frequency",
       y = "SNP Number") +
  theme_minimal()

```